orchestra/model/configuration/configuration.py

```diff
@@ -0,0 +1,321 @@
+
+        self.remotes = self._get_remotes()
+        self.binary_archives_remotes = self._get_binary_archives_remotes()
+        self.branches = self._get_branches()
+
+        self._user_paths = self.parsed_yaml.get("paths", {})
+
+        remote_heads_cache_path = os.path.join(self.orchestra_dotdir, "remote_refs_cache.json")
+        self.remote_heads_cache = RemoteHeadsCache(self, remote_heads_cache_path)
+
+        self._parse_components()
+
+    def get_build(self, comp_spec):
+        component_name, build_name = parse_component_name(comp_spec)
+        component = self.components.get(component_name)
+        if not component:
+            return None
+        if build_name:
+            build = component.builds[build_name]
+        else:
+            build = component.default_build
+        return build
+
+    def global_env(self):
+        env = OrderedDict()
+        env["ORCHESTRA_DOTDIR"] = self.orchestra_dotdir
+        env["ORCHESTRA_ROOT"] = self.orchestra_root
+        env["SOURCE_ARCHIVES"] = self.source_archives
+        env["BINARY_ARCHIVES"] = self.binary_archives_dir
+        env["SOURCES_DIR"] = self.sources_dir
+        env["BUILDS_DIR"] = self.builds_dir
+        env["TMP_ROOTS"] = self.tmproot
+        env["RPATH_PLACEHOLDER"] = "////////////////////////////////////////////////$ORCHESTRA_ROOT"
+        env["GIT_ASKPASS"] = "/bin/true"
+
+        # TODO: the order of the variables stays the same even if the user overrides an
+        #       environment variable from the config. This is convenient but we should
+        #       think if it is really what we want.
+        for env_dict in self.parsed_yaml.get("environment", {}):
+            for k, v in env_dict.items():
+                env[k] = v
+
+        path = ":".join(self.parsed_yaml.get("add_to_path", []))
+
+        for component in self.components.values():
+            for additional_path in component.add_to_path:
+                path += f":{additional_path}"
+
+        path += "${PATH:+:${PATH}}"
+        env["PATH"] = path
+
+        return env
+
+    def get_suggested_component_name(self, user_component_name):
+        best_ratio = 0
+        best_match = None
+        for component_name in self.components:
+            ratio = fuzz.ratio(user_component_name, component_name)
+            if ratio > best_ratio:
+                best_ratio = ratio
+                best_match = component_name
+        return best_match
+
+    def _parse_components(self):
+        # First pass: create the components, their builds and actions
+        for component_name, component_yaml in self.parsed_yaml["components"].items():
+            component = Component(component_name, component_yaml, self)
+            self.components[component_name] = component
+
+        # Second pass: resolve dependencies
+        for component in self.components.values():
+            component.resolve_dependencies(self)
```

Comment:
Oh this is nice. Exactly how it should look like. And you were doubting this would have been an improvement! :)